### PR TITLE
feature: upgrade provider to work with gopass 1.10

### PR DESCRIPTION
**Pull request by toabi**
Also upgrade terraform-plugin-sdk to v2.

This is my first Go project I'm working on… so I'm happy for comments about the style/code/…

It might be a v2.0 release of this provider because it is now only compatible with terraform 0.12+

As a new feature it allows users to set a YAML document as secret. 
I cleaned up the provider options which don't work anymore (refresh & store location) (so #42 will be "solved" :D)

This should also solve #41.

### CODE REVIEW

**Comment by toabi** [comment]
I made a quick release if somebody want's to test it without compiling it: https://github.com/toabi/terraform-provider-pass/releases/tag/v2.0.0-rc1

**Comment by toabi** [comment]
@jaceq Maybe you want to try this version on your repos?

**Comment by jaceq** [comment]
@toabi Very sorry about this delay, I missed your comment.
I just tested and it does work.
I tested following configurations and they all work:
- new provider with gopass 1.9.2 (binary and structure)
- new provider with gopass 1.10 (but a folder /config structure from 1.9.2 - this has worked also before)
- new provider with gopass 1.10 with re-created config + store (this has NOT worked before)

So LGTM

**Comment by jaceq** [comment]
@raphink could someone have a look at that?

**Comment by raphink** [comment]
Sorry for the lag @toabi @jaceq. I will take a look.

So far, we've been reluctant to migrate to Gopass 1.10 as lots of things have broken in this release (especially with YAML support)

File: pass/helpers.go [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+	st := provider.store
+	log.Printf("reading %s from gopass", path)
+
+	sec, err := st.Get(context.Background(), path, "") //TODO: support getting a revison via terraform?
+	if err != nil {
+		return errors.Wrapf(err, "failed to read password at %s", path)
+	}
+
+	// Retrieve all data items if keys exist
+	var keys = sec.Keys()
```
Comment (raphink) [inline pass/helpers.go:37]: It looks like `keys` is not used if `readData` is `false`, so you might want to wrap this line in `if readData`

File: pass/resource_pass_password.go [hunk 2]
```diff
@@ -24,70 +22,103 @@ func passPasswordResource() *schema.Resource {
...
+		value = "---\n" + yaml
 	} else {
-		sec := secret.New(passwd, fmt.Sprintf("---\n%s", dataYaml))
-		err = st.Set(context.Background(), path, sec)
+		var result bytes.Buffer
+		tmpl, err := template.New("secret").Parse(`{{ .Password}}
+{{ range $key, $value := .Data }}
+{{ $key }}: {{ $value }}
+{{ end }}
+`)
```
Comment (raphink) [inline pass/resource_pass_password.go:91]: Maybe it would be simpler to `strings.Join()` the elements by `:` then `\n`?

**Review by raphink — CHANGES_REQUESTED** [review]

File: pass/resource_pass_password.go [hunk 3]
```diff
@@ -24,70 +22,101 @@ func passPasswordResource() *schema.Resource {
...
+		tmpl, err := template.New("secret").Parse(`
+{{ range $key, $value := .Data }}
+{{ $key }}: {{ $value }}
+{{ end }}`)
+		err = tmpl.Execute(&result, &templateData{data})
+		if err != nil {
+			return errors.New("error when rendering secret template")
+		}
+		value += result.String()
 	}
```
Comment (raphink) [inline pass/resource_pass_password.go:96]: You could replace the template with something along the lines of:

```go
  elems := make([]string, len(data))
  for k, v := range data {
    elems = append(elems, fmt.Sprintf("%s: %s", k, v))
  }

  value := strings.Join(elems, "\n")
```

**Comment by jaceq** [comment]
@toabi any update here maybe?
I am using your fork already, but would be very cool to be able to switch to upstream ;)

File: pass/resource_pass_password.go (hunk 3 above)
Comment (raphink) [reply pass/resource_pass_password.go:96]: @toabi did you see this comment?

Comment (toabi) [reply pass/resource_pass_password.go:96]: Yeah. Sorry I didn't have much time recently to work on this more.

Comment (jaceq) [reply pass/resource_pass_password.go:96]: @toabi will you have some time or should re-fork your fork and redo this PR with changes?

Comment (toabi) [reply pass/resource_pass_password.go:96]: @jaceq You're welcome to re-fork my fork to get this freaking fix flying 🥳

**Comment by jaceq** [comment]
So I've opened new PR: https://github.com/camptocamp/terraform-provider-pass/pull/45 I guess this one can be closed in favor of the new one.

**Comment by raphink** [comment]
OK, I'm having a look at #45 then.

**Pull request closed without merging**